dvc/utils/stream.py

```diff
@@ -70,6 +70,7 @@ class HashedStreamReader(io.IOBase):
     def __init__(self, fobj):
         self.fobj = fobj
         self.md5 = hashlib.md5()
+        self.total_read = 0
```

Comment:
We should probably extend `upload_fobj` with `size` for Webdav and HTTP servers as we want to avoid chunked encoding as much as possible (`PUT` does not have good support for `Transfer-Encoding: chunked`, see https://github.com/iterative/dvc/issues/4796). 

fsspec based file object already has a size attribute which we could use to pass that to `upload_fobj`.

Reply:
Would you mind extending it a bit?

Reply:
As WebDAV uses PUT for upload, there might be old servers that don't support chunked encoding. The same applies to HTTP. So, we try to avoid chunking upload as much as possible. But to do that, we need to set the `Content-Length` header with the expected file size. 

So, I was suggesting something like following:
```python
with fs.open(...) as fd:
    stream = HashedStreamReader(fd)
    size = getattr(fd, "size", None)
    odb.fs.upload_fobj(stream, size=size)
```

But looking at it right now, it seems to have been implemented already for the `pbar` args, but we could rephrase/repurpose it.
https://github.com/iterative/dvc/blob/5fb181fcf8b58e91268a51301bd6d43052921c76/dvc/objects/stage.py#L21-L23

Reply:
Ah, I see. Regarding the usage of `fd.size`, I see that it is not required in general so that is alright but just to keep in mind I have some ideas about making it lazy so that we don't make an additional API call every time (also it is only supported in AbstractBufferedFile's, so stuff like sshfs.SSHFile won't be supporting it).

Reply:
Could you please share what you are thinking about sizes, only related to `sshfs` or in DVC itself? Anyway, I am moving with `size` stuff for now at I am considering it as broken.